osquery/events/pathset.h

```diff
@@ -11,149 +11,142 @@
- *                    treated as pattern).
  */
 template <typename PathType>
 class PathSet : private boost::noncopyable {
  public:
   void insert(const std::string& str) {
     auto pattern = str;
     replaceGlobWildcards(pattern);
-    auto vpath = PathType::createVPath(pattern);
+    auto path = PathType::createPath(std::move(pattern));
 
     WriteLock lock(mset_lock_);
-    for (auto& path : vpath) {
-      paths_.insert(std::move(path));
-    }
+    patterns_.push_back(std::move(path));
   }
 
   bool find(const std::string& str) const {
     auto path = PathType::createPath(str);
 
     ReadLock lock(mset_lock_);
-    if (paths_.find(path) != paths_.end()) {
-      return true;
+    for (const auto& pattern : patterns_) {
+      if (compare(pattern, path)) {
+        return true;
+      }
     }
     return false;
   }
 
   void clear() {
     WriteLock lock(mset_lock_);
-    paths_.clear();
+    patterns_.clear();
   }
 
   bool empty() const {
     ReadLock lock(mset_lock_);
-    return paths_.empty();
+    return patterns_.empty();
   }
 
  private:
   typedef typename PathType::Path Path;
   typedef typename PathType::Compare Compare;
-  std::multiset<Path, Compare> paths_;
+  std::vector<Path> patterns_;
   mutable Mutex mset_lock_;
+  Compare compare;
 };
 
 class patternedPath {
  public:
   typedef boost::tokenizer<boost::char_separator<char>> tokenizer;
   typedef std::vector<std::string> Path;
-  typedef std::vector<Path> VPath;
   struct Compare {
+    bool compareStrings(boost::string_view pattern,
+                        boost::string_view str) const {
+      if (pattern[0] == '*' && pattern[pattern.size() - 1] == '*') {
+        pattern = pattern.substr(1, pattern.size() - 2);
+        if (str.size() >= pattern.size() &&
+            str.find(pattern) != boost::string_view::npos) {
+          return true;
+        }
+        return false;
+      } else if (pattern[0] == '*') {
+        pattern = pattern.substr(1);
+        if (pattern.size() <= str.size()) {
+          str = str.substr(str.size() - pattern.size());
+        } else {
+          return false;
+        }
+      } else if (pattern[pattern.size() - 1] == '*') {
+        pattern = pattern.substr(0, pattern.size() - 1);
+        if (pattern.size() <= str.size()) {
+          str = str.substr(0, pattern.size());
+        } else {
+          return false;
+        }
+      }
+
+      return (pattern == str);
+    }
+
     bool operator()(const Path& lhs, const Path& rhs) const {
-      size_t psize = (lhs.size() < rhs.size()) ? lhs.size() : rhs.size();
+      auto psize = std::min(lhs.size(), rhs.size());
       unsigned ndx;
       for (ndx = 0; ndx < psize; ++ndx) {
-        if (lhs[ndx] == "**" || rhs[ndx] == "**") {
-          return false;
+        if (lhs[ndx] == "**") {
+          return true;
         }
 
-        if (lhs[ndx] == "*" || rhs[ndx] == "*") {
+        if (lhs[ndx] == "*") {
           continue;
```

Comment:
This can not happen, Since lhs[ndx] is string not an element of a string. Here string comparison is happening.  e.g.
/xyz/*/uvw 
tokenizer will break this in {"xyz", "*", "uvw"}.

Reply:
lhs[0] is "a"
lhs[0] != "**" 
lhs[0] != "*"
lhs[0] == rhs[0] == "a"
match (continue)
lhs[1] is "b*c"
lhs[1] != "**" 
lhs[1] != "*"
lhs[1] != rhs[1]  
no match (return)

Result  
lhs = {"a", "b*c", "d"}
rhs = {"a", "k", "d"}
dont match

Reply:
Oh, I see   your  point now. Thanks